src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h

```diff
@@ -0,0 +1,162 @@
+
+#include <string>
+#include <set>
+#include <memory>
+
+#include "veins/modules/mobility/traci/TraCIBuffer.h"
+#include "veins/modules/mobility/traci/TraCICommandInterface.h"
+#include "veins/modules/mobility/traci/TraCIConnection.h"
+
+
+namespace veins {
+namespace TraCISubscriptionManagement {
+
+/**
+ * @class SubscriptionManagerBase
+ *
+ * This class provides some common functionality for a subscription manager. It stores
+ * TraCIConnection, TraCICommandInterface and a set containing subscribed ids.
+ *
+ * It also enforces you to provide an update() method in your subscription manager.
+ *
+ * @author Nico Daßler <dassler@hm.edu>
+ */
+class SubscriptionManagerBase {
+
+public:
+    /**
+     * Constructor.
+     */
+    SubscriptionManagerBase();
+
+    /**
+     * Default destructor.
+     */
+    virtual ~SubscriptionManagerBase() = default;
+
+    /**
+     * Each subscription manager has to implement this method. The given buffer
+     * is assumed to have the first few bytes removed. These bytes represent
+     * the first few fields of an answer to a simulation step request or the
+     * request to subscribe variable request.
+     * The fields that should be removed are:
+     *      - the number of subscription results
+     *      - the response command length
+     *      - the response command length extended
+     *      - the id of the response commmand
+     *
+     * This method is expected to change the results to api methods of this manager.
+     *
+     * @param buffer the TraCIBuffer containing the subscription response.
+     *
+     * @return if there was an id_list was received in the buffer (not all subscription manager will use
+     * this. the result is undefined for them).
+     */
+    virtual bool update(TraCIBuffer& buffer) = 0;
+
+    /**
+     * This initializes this manager with the necessary connection and command interface
+     * to perform requests to the TraCI server.
+     *
+     * Important: Always do this before using a subscription manager.
+     *
+     * Note: When overriding the this method from a child class make sure to call the base
+     * class implementation to set mConnection and mCommandInterface variables.
+     *
+     * @param connection to access traci server (for example: to perform subscriptions).
+     * @param commandInterface to access traci (for example: direct request to get vehicle id_list)
+     */
+    virtual void initialize(std::shared_ptr<TraCIConnection> connection, std::shared_ptr<TraCICommandInterface> commandInterface);
+
+    /**
+     * Checks if this id is subscribed to by this manager.
+     *
+     * @param id to check.
+     *
+     * @return true if the given id is subscribed. False if not.
+     */
+    bool isSubscribed(std::string id);
+
+    /**
+     * Subscribe to the given id.
+     *
+     * @param id to subscribe to.
+     *
+     * @return true if adding was successful.
+     */
+    bool addToSubscribed(std::string id);
```

Comment:
Name:
I did not call it subscribe because there is no subscription performed here. The id is only added to a set of subscribed ids. (The comment above the method is misleading though. I will change that.)

Exception + return value:
I will throw an exception when a duplicate got added.

Reply:
All clear about the error handling.

For the name and semantics: Is there a way to register some object for subscription then? Or can that only be done via `ID_LIST` results?

Reply:
That depends on the manager. The methods defined in the base class of each manager are just for tracking which ids have been subscribed and are there to reduce the amount of duplicate code I wrote.

The different managers implement their own methods to subscribe to a specific id (for example persons or vehicles) at the traci server. For example: https://github.com/truenicfel/veins/blob/subsRework/src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc#L231
*(This is from my fork of veins)*

The `ID_LIST` subscription tells me which ids of cars are currently active. The container in the manager base class will get updated with these infos. The newly appeared ids will then be fed
into a method similar to the one linked above.